core/interop/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+//! Interop utilities between Boa and its host.
+use std::cell::RefCell;
+
+use boa_engine::module::SyntheticModuleInitializer;
+use boa_engine::{Context, JsString, JsValue, Module, NativeFunction};
+
+pub mod loaders;
+
+/// A trait to convert a type into a JS module.
+pub trait IntoJsModule {
+    /// Converts the type into a JS module.
+    fn into_js_module(self, context: &mut Context) -> Module;
+}
+
+impl<T: IntoIterator<Item = (JsString, NativeFunction)> + Clone> IntoJsModule for T {
+    fn into_js_module(self, context: &mut Context) -> Module {
+        let (names, fns): (Vec<_>, Vec<_>) = self.into_iter().unzip();
+        let exports = names.clone();
+
+        Module::synthetic(
+            exports.as_slice(),
+            unsafe {
+                SyntheticModuleInitializer::from_closure(move |module, context| {
+                    for (name, f) in names.iter().zip(fns.iter()) {
+                        module
+                            .set_export(name, f.clone().to_js_function(context.realm()).into())?;
+                    }
+                    Ok(())
+                })
+            },
+            None,
+            context,
+        )
+    }
+}
+
+/// A trait to convert a type into a JS function.
+pub trait IntoJsFunction {
+    /// Converts the type into a JS function.
+    fn into_js_function(self, context: &mut Context) -> NativeFunction;
+}
+
+impl<T: FnMut() + 'static> IntoJsFunction for T {
+    fn into_js_function(self, _context: &mut Context) -> NativeFunction {
+        let s = RefCell::new(self);
+
+        unsafe {
+            NativeFunction::from_closure(move |_, _, _| {
+                s.borrow_mut()();
+                Ok(JsValue::undefined())
+            })
+        }
+    }
+}
+
+#[test]
+#[allow(clippy::missing_panics_doc)]
+pub fn into_js_module() {
```

Comment:
thought: could this test be moved into a doc test

This crate is going to be heavily user facing, so having a test like this visible in rustdoc will probably be incredibly important.

Reply:
This test is going to change very soon. I could make it into its own spec test, but I'll setup doc tests when I add the next layer of abstraction if you don't mind. Basically I want to add Traits that make this around ~15 lines instead of the monstrosity of boilerplate it currently is.

Basically it's very ugly and I'd rather have doctest and examples that are higher in abstraction (for the sake of the reader). WDYT?

Reply:
For reference, I've created a PR that's based on this PR with the next step I want to work on: https://github.com/boa-dev/boa/pull/3773

There is a doctest for the `IntoJsFunction`, but still none for `IntoJsModule`.